src/server/json.ts

```diff
@@ -0,0 +1,53 @@
+import * as t from "io-ts"
+import { PathReporter } from "io-ts/lib/PathReporter"
+import { RequestHandler, Request, Response } from "express"
+
+const isProd = () => process.env.NODE_ENV === "production"
+
+export class NotFoundError {
+  constructor(readonly message: string) {}
+}
+
+export class ValidationError {
+  constructor(readonly errors: string[]) {}
```

Comment:
Mites tää toimii, toi errors ei taida tallentua mihinkään? Näkyy ilmeisesti jos tulee unexpected error mutta saako sen luettua näin?

```
var error = new ValidationError("INVALID")
 error.errors
```

Reply:
Joo, se menee automaagisesti `this.errors` propertyks kun parametrille antaa public/private/readonly määreen constructorissa.

Eli vastaa täysin tätä:

```ts
class ValidationError {
  readonly errors: string[]
  constructor(errors: string[]) {
    this.errors = errors
  }
}
```

https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties

Reply:
Näppärää erityisesti jos haluu lähinnä antaa jollekin datarakenteelle nimen ja tyypityksen. Jossain kielissä on vastaava konsepti termillä `case class`.

Esim. Scalassa vastaava olis näinki simppeli:

```scala
case class ValidationError(errors: List[String])

throw ValidationError(List("INVALID"))
```